### DISCUSSION TITLE
Fixing Jupyter Based Demos

### DISCUSSION THREAD

**Discussion by CSSFrancis** [hyperspy/hyperspy#3432] [Ideas]
Supporting jupyter based demos is quite difficult:

Currently the [pyxem-demos](https://github.com/pyxem/pyxem-demos) are failing and hard to update. At the same time real examples with real data is very useful. 

**We have demos from different workshops**

- https://github.com/TEM-Gemini-Centre/nordtemhub_4DSTEMworkshop_2024
- https://github.com/hyperspy/EMC2024_HyperSpy_Workshop
- https://github.com/ePSIC-DLS/Hyperspy_Workshop_2024
- https://github.com/the-microanalysis-society/2023-IUMAS-workshop
- https://github.com/LumiSpy/eBEAM2024-Tutorial

**Workbooks Associated with Different Papers**
- https://github.com/CSSFrancis/4d_stem_clustering
- https://github.com/din14970/pyxem_template_matching_workflows
- https://github.com/emichr/PhaseMappingPaper
- https://github.com/hakonanes/p-texture-al-mn-alloys
- https://github.com/hakonanes/correlated-grains-particles-workflow
- https://github.com/elisathr/SPED-phase-mapping
- ... I'm probably forgetting some.

Proposal:
---------

We could easily build a list of repos using nbsphinx and serve it from the hyperspy website.  Then each repo would just have to be tagged with a hyperspy/ pyxem/kikchipy version.  Some of the repos we could autorun and check to see if they support newer versions but for the most part it would just be static versions of the notebooks.  Then we could link to the repos in pyxem, hyperspy, kikuchipy, lumispy etc. 

@magnunor @pc494 @hakonanes @jlaehne @ericpre Thoughts?

**Comment by ericpre** [comment]
This is true that for users it can be quite confusing and it should be streamlined. This has been bugging me for a while but with the recent improvement in documentation, we may be now in a place to tackle this issue!
I would suggest to make the distinction between 3 types of repositories; here I am using hyperspy as an example, but the same would apply to other libraries:
1. Tutorials:
   - maintained
   - expected to work with most recent version
   - example: https://github.com/hyperspy/hyperspy-demos
2. Workshop:
   - usually a combination, re-use of various tutorials from point 1. matching a specific agenda, re-use of previous workshop repositories
   - only used for limited period of time for a specific period of time
   - not maintained
   - example: https://github.com/ePSIC-DLS/Hyperspy_Workshop_2024
3. Papers:
   - notebook to reproduce published results
   - not maintained

To improve the situation, the following actions could be done (and recommended practise), some are from [@CSSFrancis's suggestion from above](https://github.com/hyperspy/hyperspy/discussions/3432#discussion-7142215):
1. Tutorials:
    - Add CI to make sure they run with latest version
    - Push to github pages a html version of the notebooks, add links to relevant documentation (https://hyperspy.org/, https://hyperspy.org/hyperspy-doc/current, etc.)
    - ideally the html version should have the links to binder, jupyterlite, etc.
    - "backport" improvement from workshop into tutorials
2. Workshop:
    - archive the repositories after they have been used with a note specifying the version required explaining that they are not maintained and that there is no guaranty that they will work with recent version
3. Papers:
    - archive - if not expected to be maintained, add instruction to setup environment with compatible libraries?
    - recommend adding tags (for example, `paper` and name of relevant libraries) to discover these notebooks, for example using https://github.com/search?q=topic%3Atutorial+topic%3Ahyperspy&type=repositories but with `paper` tag instead of `tutorial`

I find the name of `hyperspy-demos` a bit misleading and maybe we should rename it to `hyperspy-tutorial`?

**Reply by CSSFrancis** [reply]
@ericpre I think that is a good suggestion.  My one addition would be that it would be nice to have all of the tutorials served from one website with tags for each one. I saw a really nice example of this the otherday and can't remember where....

- https://learn.astropy.org
- https://examples.holoviz.org/gallery/index.html
- http://gallery.pangeo.io


The easiest way to do this would be to have a simple standard json file for each entry:

- Link to Repo 
- DOI/ DOIs Associated with Repo (optional)
- Data Asscoated with Repo (optional)
- Hyperspy/ extension versions --> also used to tag if related to an extension
- Tutorial vs Workshop vs Paper

Then down stream in pyxem I would probably just point the Tutorials page towards the hyperspy repo with the "pyxem" tag selected.  I think that would help people coming from pyxem to look at hyperspy tutorials and the other way as well.

**Reply by ericpre** [reply]
Yes, these examples are very useful indeed. It seems a bit more involved to setup but as you said most likely worth it to help users navigate the different tutorial repository.

**Reply by jlaehne** [reply]
> Workshop:
> 
>     * archive the repositories after they have been used with a note specifying the version required explaining that they are not maintained and that there is no guaranty that they will work with recent version

For tutorial repositories, we could maybe additionally fix the versions in the `environment.yml` file so that they are at least continue running on binder?

**Comment by jlaehne** [comment]
In `lumispy-demos` I just added a Readme to the `Workshops` folder to warn that such notebooks are not maintained: https://github.com/LumiSpy/lumispy-demos/pull/21
